### ISSUE TITLE
Support for Prince specific styling

### ISSUE THREAD

**Issue by jenstroeger** [stylelint/stylelint#7284]
### What is the problem you're trying to solve?

Styling documents for print using [Prince styles](https://www.princexml.com/doc/styling/) goes beyond CSS for web pages using custom functions, properties, selectors, and more.

I’m currently exploring `stylelint` and I was wondering if there’s any support available as a plugin, perhaps?

### What solution would you like to see?

Uncertain at this point… a plugin with custom rules, perhaps?

**Comment by ybiquitous** [comment]
@jenstroeger Yes, you're correct. I'm unsure about Prince styles, but I you should create a plugin or ruleset for the styles.

If you publish a new package (plugin/config), please consider adding it to [Awesome Stylelint](https://github.com/stylelint/awesome-stylelint).

**Comment by jenstroeger** [comment]
Thanks @ybiquitous!

Would a plugin be able to check Prince-specific properties (e.g. any of the [`-prince-*` properties](https://www.princexml.com/doc/css-props/)) or function expressions (e.g.  [`prince-*` function expressions](https://www.princexml.com/doc/css-functions/)) and so forth? 

Also, I’m new to `stylelint` so would [Writing Plugins](https://github.com/stylelint/stylelint/blob/main/docs/developer-guide/plugins.md) be the place to get started? Is there perhaps an empty plugin template from which I can start? Any other “getting started” tips I ought to know about?

**Comment by ybiquitous** [comment]
@jenstroeger If you want just to restrict the use of properties, functions, etc., you can use the built-in rules (`*-allowed-list`/`*-disallowed-list`) (see also the [doc](https://stylelint.io/user-guide/rules#allowed-disallowed--required)). And then, you can also publish such a ruleset as an npm package like [stylelint-config-recommended](https://www.npmjs.com/package/stylelint-config-recommended).

If the built-in rules are insufficient in your use case, you can write a plugin. The following guide could be helpful:
- https://stylelint.io/developer-guide/rules
- https://stylelint.io/developer-guide/plugins

The [plugins in Awesome Stylelint](https://github.com/stylelint/awesome-stylelint#plugins) might be helpful as a reference, too.

**Comment by jenstroeger** [comment]
I think it’s a bit more than that — I’d like to address the following errors, all of which I believe are related to Prince specifics:
```
✖  Unexpected unknown function "prince-opentype"                        function-no-unknown
✖  Unexpected unknown property "prince-text-replace"                    property-no-unknown
✖  Unexpected unknown at-rule "@footnotes"                              at-rule-no-unknown
✖  Unexpected unknown type selector "frontmatter"                       selector-type-no-unknown
```
The type selector _may_ be now outdated with Prince, I need to check.

Also, `stylelint` doesn’t know the values for custom properties, e.g.
```css
part {
  prince-pdf-tag-type: Part;  /* ✖  Expected "Part" to be "part" */
  page: page-nofolio;
}
```
I think for now this would be a good start to handle, but I need to rely on your guidance to get started: should I configure or create a plugin? What would be more flexible going forward?

And yes, absolutely, I’d be delighted to make this available to the community! (I’m somewhat surprised that such an effort doesn’t exist yet, though 😳)

**Comment by ybiquitous** [comment]
A more flexible and extensible way is to create a plugin, but there is a downside: knowledge about AST (and PostCSS) is necessary. You can start either way.

**Comment by jenstroeger** [comment]
I’m familiar with ASTs (I assume you mean [Abstract Syntax Trees](https://en.wikipedia.org/wiki/Abstract_syntax_tree)) but I’ve not used PostCSS before — [this one](https://github.com/postcss/postcss)?

**Comment by ybiquitous** [comment]
Correct. Stylelint uses the PostCSS's AST. You can see the AST on https://astexplorer.net/ 👍🏼

**Comment by ybiquitous** [comment]
I'm closing this issue because this repository won't touch on specific plugins. Feel free to let us know if you create a plugin (or sharable config).
